File: 869.reordered-power-of-2.py
#
# @lc app=leetcode id=869 lang=python3
#
# [869] Reordered Power of 2
#

# @lc code=start
class Solution:
    def reorderedPowerOf2(self, n: int) -> bool:
        n_str = str(n)
        n_str = ''.join(sorted(n_str))
        width = len(n_str)
        power2 = 1
        while len(power2_str := str(power2)) <= width:
            power2 <<= 1
            if len(power2_str) < width:
                continue
            power2_str = ''.join(sorted(power2_str))
            if power2_str == n_str:
                return True
        return False
# @lc code=end

File: 869.reordered-power-of-2.cpp
/*
 * @lc app=leetcode id=869 lang=cpp
 *
 * [869] Reordered Power of 2
 */
#include <vector>
#include <algorithm>
using namespace std;
// @lc code=start
class Solution
{
public:
    bool reorderedPowerOf2(int n)
    {
        string n_str = to_string(n);
        int width = n_str.size();
        sort(n_str.begin(), n_str.end());
        long power2 = 1;
        string power2_str;
        while ((power2_str = to_string(power2)).size() <= width)
        {
            power2 <<= 1;
            if (power2_str.size() < width)
            {
                continue;
            }
            sort(power2_str.begin(), power2_str.end());
            if (power2_str == n_str)
            {
                return true;
            }
        }
        return false;
    }
};
// @lc code=end
